File: lib/File.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <dirent.h>

typedef struct {
	FILE *fp;
	int size;
	char *name;
} file_t;

struct files_t {
	file_t **file;
	int cur_files;
} files;

int File_size(char *name) {
	FILE *fp = fopen(name, "rb");
	fseek(fp, 0, SEEK_END);
	int sz = ftell(fp);
	fseek(fp, 0, SEEK_SET);
	fclose(fp);
	return sz;
}
int File_size_by_id(int id) {
	if(id < files.cur_files)
		return files.file[id]->size;
	return -1;
}
int File_open(char *name, char *mode) {
	file_t *f = malloc(sizeof(file_t));
	f->fp = fopen(name, mode);
	if(!f->fp) { free(f); return 0; }
	f->size = File_size(name);
	f->name = name;
	if(files.file == NULL) { // initialize
		files.file = calloc(2, sizeof(file_t));
		files.cur_files = 1;
	} else files.file = realloc(files.file, (1+files.cur_files)*sizeof(file_t));
	files.file[files.cur_files++] = f;
	return files.cur_files - 1;
}
char *File_read(int id) {
	if(id < files.cur_files) {
		file_t *f = files.file[id];
		char *str = (char *)calloc(f->size+1, 1);
		fread(str, f->size, 1, f->fp);
		return str;
	} return NULL;
}
void File_write(int id, char *str) {
	if(id < files.cur_files) 
		fwrite(str, strlen(str), 1, files.file[id]->fp);
}
DIR *dir;
char *File_dir_list(char *cur_dir_name) {
	struct dirent *dp;
	char **dirs = calloc(1, sizeof(char *));
	int dir_count = 0;

	if(dir == NULL) {
		dir = opendir(cur_dir_name);
		if(dir == NULL) return NULL;
	}

	dp = readdir(dir);
	if(dp == NULL) { 
		closedir(dir); dir = NULL; 
		return NULL; 
	}
	return dp->d_name;
}
void File_close(int id) {
	if(id < files.cur_files) {
		file_t *f = files.file[id];
		fclose(f->fp);
		free(f);
	}
}

File: lib/File.rb
lib File
	proto File_open(name:string mode:string)| file_open
	proto File_write(id str:string) | file_write
	proto File_read(id):string | file_read
	proto File_close(id) | file_close
	proto File_dir_list(dir:string):string
	proto File_size_by_id(id)
	proto File_ctime(id):TIME
end

struct FILE
	id
	name:string
	use?
end

module File
	def new:FILE
		new FILE
	end
	def open name:string mode:string
		file_open name mode
	end
	def open name:string
		file_open name "r"
	end
	def write id str:string
		file_write id str
	end
	def read id:int :string
		file_read id 
	end
	def read name:string :string
		id = open name
		if id == 0
			return "error: no such file or directory"
		end
		content = read id
		file_close id
		content
	end
	def dir_list_one(dir:string):string
		File_dir_list dir
	end
	def dir_list(dir:string):string[]
		list = []:string
		while true
			dir = File::dir_list_one "."
			if dir == nil
				break
			end
			list += "" + dir
		end
		list
	end
	def close id
		file_close id
	end
	def size id
		File_size_by_id id 
	end
	def ctime(id):TIME
		File_ctime id
	end

	def exist?(name:string)
		a = File::open name
		if a == 0
			false
		else
			File::close a
			true
		end
	end
end


def open(ref file:FILE, name:string, mode:string):FILE
	if file.use?
		file.close()
	end
	file.id = File::open name, mode
	file.name = name
	file.use? = true
	if file.id == 0
		<FILE>nil
	else
		file
	end
end
def open(ref file:FILE, name:string):FILE
	open(file, name, "r")
end
def read(file:FILE):string
	File::read file.id
end
def write(file:FILE, content:string)
	File::write file.id, content
end
def close(ref file:FILE)
	File::close file.id
	file.use? = false
end
def size(file:FILE)
	File::size file.id
end
def ctime(file:FILE):TIME
	File::ctime file.id
end
